docs/src/03_configuration.md
# Configuration

`nut_webgui` can be configured via CLI args, environment variables, or a config
file. All configuration options are merged into a single unified config based on
their priority.

Config priority order, highest to lowest:
1. [CLI args](./03_01_cli_args.md)
2. [Environment variables](./03_02_env_variables.md)
3. [config.toml file](./03_03_config_file.md)


docs/src/03_03_config_file.md
# config.toml

`config.toml` has the least priority, but it is recommended to use the config
file as a baseline configuration and use environment variables and command-line
arguments to override settings when needed.

```toml
version = "1"
log_level = "info"
default_theme = "tokyo-night"

[http_server]
base_path = "/"
listen = "0.0.0.0"
port = 9000
worker_count = 8

[upsd.default]
username = "admin"
password = "where an old man of Aran goes around and around"
address = "localhost"
port = 3493
max_connection = 4
poll_freq = 30
poll_interval = 2
tls_mode = "disable"

[upsd.reactor]
address = "10.0.12.10"
username = "observer"
password = "AbsoluteSecurity"

[auth]
users_file = "/etc/nut_webgui/users.toml"
allow_anonymous_metrics = false
```


CHANGELOG
## v0.10.3 - September 02, 2026

### Fixed
- Previous `install.sh` does not work on Alpine Linux or similar distributions
  due to some incompatible command calls. The script now tries to use
  alternatives *(e.g. `doas` instead of `sudo`)* and uses common option flags.
- Auto-generated config files inside the container are now owned by the provided
  `UID` instead of the root user.
- Default configuration files are not created when users explicitly set
  environment variables to the default paths.

### Added
- `--server-key-file` option for setting the server key from a file.
- Self-contained RISC-V build with musl.
- runit and OpenRC support for binary installations.

### Changed
- Tar packages now contain all necessary config templates, man docs, and basic
  service files *(runit, openrc, systemd)*.
- Due to file upload limits, SHA256 checksums are bundled in a `MANIFEST` file.

### Removed
- PKGBUILD file for Arch-based systems is no longer distributed with new
  releases.

## v0.10.2 - June 29, 2026

### Fixed
- UPS devices without a serial number produce an invalid topology graph.
- The AC power symbol can be misaligned on some fonts. It now uses a custom SVG
  icon.

## v0.10.1 - June 26, 2026

### Added
- Option to disable authentication on the `/metrics` endpoint.

### Changed
- If UPS drivers don't report ups.power and ups.realpower, the metric exporter
  tries to derive `nutwg_ups_power_voltamps` and `nutwg_ups_realpower_watts`
  from ups.load and nominal values.

### Removed
- Removed `nutwg_ups_input_frequency_hertz` and `nutwg_ups_input_voltage_volts`
  metrics.

## v0.10.0 - June 06, 2026

### Added
- UID/GID environment variables to change the running user.
- `/metrics` endpoint with OpenMetrics 1.0 format.

### Changed
- Some npm packages used for bundling JS had post-install scripts. Those
  packages were replaced by alternatives. (Bundling "modern" JS almost makes you
  wish for programming with punch cards)

## v0.9.2 - May 06, 2026

### Fixed
- Client assets bundler fails to build when either npm or pnpm is missing on the
  host system instead of selecting one of them.
- Compiling nut_webgui on FreeBSD fails due to minor differences in `sockaddr`
  types.

### Changed
- Failure on listing writable variables, instant commands, and monitoring
  clients are now considered as soft-fail. This change allows using
  non-compliant NUT servers to work with nut_webgui.
- NUT TCP connections now reuse their buffers to minimize memory allocation.

## v0.9.1 - April 12, 2026

### Fixed
- rustls's openssl-prober no longer checks the /usr/share/ssl directory, which
  breaks certificates mounted via the UPSD_ROOT_CA environment variable.

## v0.9.0 - February 07, 2026

### Added
- New events API based on web sockets.
- Mimalloc set as main memory allocator for x86-64 and arm64 based builds.
- Worker-count configuration options to change core count used by nut_webgui.
  With Mimalloc, this setting plays an important role in tweaking application
  memory usage.

### Changed
- Glorious potato symbol removed from README. They used to be just a
  funny-cringe addition, but nowadays they symbolize slop.

### Fixed
- UPS device stops updating and only shows previous data when UPSD driver fails
  with DRIVER-NOT-CONNECT.
- Failed UPS device status is now set as `NOCONN` (no connection).

### Deprecated
- Old config file format used in earlier versions is finally deprecated (v0.5.0
  to v0.6.0).

## v0.8.1 - January 18, 2026

### Fixed
- Data and electrical edges visually overlap when they connect to the same nodes
  on the topology page.

### Changed
- Last synchronized date now shows the localized time (based on the browser's
  locale) instead of UTC only.
- Added an updated UI indicator for when network connectivity is lost.

## v0.8.0 - December 18, 2025

### Added
- Reverse DNS lookup to find and display host names of the attached client IPs.
- TOML v1.1.0 support.

### Changed
- **(BREAKING)** The `attached` field in the device response now returns an
  array of objects instead of strings in the JSON Data API. This is a minor but
  breaking change.

### Fixed
- Topology graph nodes sometimes disappears after a re-render.

## v0.7.1 - December 8, 2025

### Added
- New topology graph page to display all linked NUT servers, clients, and UPS
  devices.

### Fixed
- Device sync background task uses `poll_interval` instead of `poll_freq`, which
  unnecessarily increases background noise.

### Changed
- UI content maximum width increased from 1584px to 2048px for better
  wide-screen monitor view.
- Home page template struct is optimized. At this point, it's a nonsense tier
  latency improvement, but why not.

## v0.7.0 - November 4, 2025

### Added
- Multiple NUT server support.
- New system page to display server, application, and user informations.
- New connection page to display upsd connections and their configurations.
- New data API endpoints for reading namespace/connection details.
- New probe endpoints for checking individual NUT server status.
- Config file version field for maintaining better backward compability in the
  future.

### Changed
- `NUTWG__UPSD__*` environment variables configure NUT server with `default`
  namespace. This allows prior nut_webgui configurations works without any change.
- The config TOML now supports multiple NUT server configurations.
- **(BREAKING)** JSON Data API endpoints are changed due to new namespace
  support. Review the OpenAPI spec before updating if you rely on `v0.6.0` API.
  This changes only affects the `/api` endpoints.

### Removed
- Brotli compression support has been temporarily removed due to excessively
  high memory usage compared to deflate and gzip compressions.

### Deprecated
- The old config TOML format (without a version field and using a single upsd
  configuration table) is deprecated. No immediate action is required for
  `v0.7.0`, but it is planned to be removed in the `v0.8.0` release.

## v0.6.0 - September 10, 2025

nut_webgui/src/config.rs
use self::error::ConfigError;
use self::utils::rand_server_key_256bit;
use self::{tls_mode::TlsMode, uri_path::UriPath};
use core::net::{IpAddr, Ipv4Addr};
use std::{collections::HashMap, num::NonZeroUsize, path::PathBuf};
use tracing::level_filters::LevelFilter;

mod utils;

pub mod cfg_arg;
pub mod cfg_env;
pub mod cfg_fallback;
pub mod cfg_toml;
pub mod cfg_user;
pub mod error;
pub mod tls_mode;
pub mod uri_path;

pub const DEFAULT_UPSD_KEY: &str = "default";

pub trait ConfigLayer {
  fn apply_layer(self, config: ServerConfig) -> Result<ServerConfig, ConfigError>;
}

pub struct ServerConfig {
  /// Additional config file path
  pub config_file: Option<PathBuf>,

  /// Web GUI default theme
  pub default_theme: Option<Box<str>>,

  /// Logging level
  pub log_level: LevelFilter,

  /// Server instance's private sign key
  pub server_key: Box<[u8]>,

  /// HTTP server configurations
  pub http_server: HttpServerConfig,

  /// UPSD connection configurations
  pub upsd: HashMap<Box<str>, UpsdConfig>,

  /// Authentication scheme configurations
  pub auth: AuthConfig,
}

#[derive(Debug)]
pub struct HttpServerConfig {
  pub listen: IpAddr,
  pub port: u16,
  pub base_path: UriPath,
  pub worker_count: Option<NonZeroUsize>,
}

#[derive(Clone)]
pub struct UpsdConfig {
  /// Poll frequency in seconds for less critical parameters
  pub poll_freq: u64,

  /// Poll interval in seconds for ups status
  pub poll_interval: u64,

  /// UPSD TCP address
  /// It can be hostname, IPv4, or IPv6
  pub addr: Box<str>,

  /// UPSD TCP port
  pub port: u16,

  /// UPSD username
  pub user: Option<Box<str>>,

  /// UPSD password
  pub pass: Option<Box<str>>,

  /// Maximum allowed connection limit aka pool size
  pub max_conn: NonZeroUsize,

  /// UPSD starts with TLS
  pub tls_mode: TlsMode,
}

#[derive(Debug)]
pub struct AuthConfig {
  pub users_file: Option<PathBuf>,
  pub allow_anonymous_metrics: bool,
}

impl AuthConfig {
  pub const fn is_enabled(&self) -> bool {
    self.users_file.is_some()
  }
}

impl UpsdConfig {
  pub fn get_socket_addr(&self) -> String {
    format!("{address}:{port}", address = self.addr, port = self.port)
  }
}

impl HttpServerConfig {
  pub fn get_listen_addr(&self) -> String {
    format!("{ip}:{port}", ip = self.listen, port = self.port)
  }
}

impl Default for AuthConfig {
  fn default() -> Self {
    Self {
      users_file: None,
      allow_anonymous_metrics: false,
    }
  }
}

impl Default for HttpServerConfig {
  fn default() -> Self {
    Self {
      base_path: UriPath::default(),
      listen: Ipv4Addr::UNSPECIFIED.into(),
      port: 9000,
      worker_count: None,
    }
  }
}

impl Default for UpsdConfig {
  fn default() -> Self {
    Self {
      addr: "localhost".into(),
      max_conn: NonZeroUsize::new(4).expect("static non-zero max_conn is provided as default"),
      pass: None,
      poll_freq: 30,
      poll_interval: 2,
      port: 3493,
      tls_mode: TlsMode::Disable,
      user: None,
    }
  }
}

impl Default for ServerConfig {
  fn default() -> Self {
    Self {
      config_file: None,
      default_theme: None,
      http_server: Default::default(),
      log_level: LevelFilter::INFO,
      server_key: rand_server_key_256bit(),
      upsd: Default::default(),
      auth: Default::default(),
    }
  }
}

impl ServerConfig {
  #[inline]
  pub fn new() -> Self {
    Self::default()
  }
}

impl ServerConfig {
  #[inline]
  pub fn layer<L>(self, layer: L) -> Result<Self, ConfigError>
  where
    L: ConfigLayer,
  {
    layer.apply_layer(self)
  }
}

impl core::fmt::Debug for UpsdConfig {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.debug_struct("UpsdConfig")
      .field("pass", &self.pass.as_ref().map(|_| "******"))
      .field("user", &self.user.as_ref().map(|_| "******"))
      .field("addr", &self.addr)
      .field("port", &self.port)
      .field("poll_freq", &self.poll_freq)
      .field("poll_interval", &self.poll_interval)
      .field("max_conn", &self.max_conn)
      .field("tls_mode", &self.tls_mode)
      .finish()
  }
}

impl core::fmt::Debug for ServerConfig {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.debug_struct("ServerConfig")
      .field("config_file", &self.config_file)
      .field("default_theme", &self.default_theme)
      .field("log_level", &self.log_level)
      .field("server_key", &"******")
      .field("http_server", &self.http_server)
      .field("upsd", &self.upsd)
      .field("auth", &self.auth)
      .finish()
  }
}
